docs/writing-tests/making-a-testing-plan.md

```diff
@@ -0,0 +1,504 @@
+to parse--should we verify them all in the tests for `document.querySelector`?
+What about `document.querySelectorAll`? Should we test them all there, too?
+
+The answers depend on the current state of the test suite: whether or not tests
+for selector parsing exist and where they are located. That's why it's best to
+confer with the people who are maintaining the tests.
+
+### Avoid excessive breadth
+
+Many algorithms operate on input that takes the form of a scalar value. When
+the set of values is finite, it can be tempting to test them all exhaustively.
+When the set is very large, test authors can reduce repetition by defining
+tests programmatically in loops.
+
+Using advanced control flow techniques to dynamically generate tests has a
+number of drawbacks. It makes the test suite more difficult to understand since
+readers have to mentally "unwind" the iteration to determine what is actually
+being verified. It also increases the risk of bugs in the tests. These bugs may
+not be obvious--they may not cause failures, and they may exercise fewer cases
+than intended.
+
+Generally speaking, such exhaustive approaches are unlikely to catch more bugs
+than a handful of carefully-chosen test cases. Although the risks of dynamic
+test generation may be tolerable in some specific cases, it's usually best to
+select the most interesting edge cases and move on.
+
+*Example* We can see this consideration in the very first step of the
+`Response` constructor from [the Fetch
+standard](https://fetch.spec.whatwg.org/)
+
+> The `Response`(*body*, *init*) constructor, when invoked, must run these
+> steps:
+>
+> 1. If *init*["`status`"] is not in the range `200` to `599`, inclusive, then
+>    [throw](https://heycam.github.io/webidl/#dfn-throw) a `RangeError`.
+>
+> [...]
+
+This function accepts exactly 400 values for the "status." With [WPT's
+testharness.js](./testharness), it's easy to dynamically create one test for
+each value. Unless we have reason to believe that a browser may exhibit
+drastically different behavior for any of those values (e.g. correctly
+accepting `546` but incorrectly rejecting `547`), then the complexity of
+testing those cases probably isn't warranted.
+
+Instead, focus on writing declarative tests for specific values which are novel
+in the context of the algorithm. For ranges like in this example, testing the
+boundaries is a good idea. `200` and `599` should not produce an error while
+`199` and `600` should produce an error. Feel free to use what you know about
+the feature to choose additional values. In this case, HTTP response status
+codes are classified by the "hundred" order of magnitude, so we might also want
+to test a "3xx" value and a "4xx" value.
+
+## Assessing coverage
+
+It's very likely that WPT already has some tests for the feature (or at least
+the specification) that you're interesting in testing. In that case, you'll
+have to study what's already been done before starting to write new tests.
```

Comment:
"study" seems to suggest a larger effort than is reasonable to ask for. I would "check" or "look for" tests that exist, but not spend more than 5-10 minutes to avoid test duplication.

Reply:
On a meta-level, is this something that we've discussed or seen as a problem? I know as a spec author and browser implementer, I have on multiple occasions worried that I'm adding duplicate tests to WPT because finding existing tests is **so damn hard**. But maybe we've already agreed that one should just learn to love the quantum duplication?

Reply:
"learn" sounds like an OK middle ground here.

I do think it's good to spend some time looking for existing tests, but think there are quickly diminishing returns in continuing to look when you don't find something. And have two tests you think are duplicates, there's a non-trivial risk you're wrong and they're both providing some value. Since the effort to deduplicate well is very high and the cost of duplication not very high, I don't tend to focus on it much.